Mathlib/Topology/DiscreteSubset.lean

```diff
@@ -0,0 +1,68 @@
+/-
+Copyright (c) 2023 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash, Bhavik Mehta
+-/
+import Mathlib.Topology.SubsetProperties
+import Mathlib.Topology.Separation
+
+/-!
+# Discrete subsets of topological spaces
+
+Given a topological space `X` together with a subset `s ⊆ X`, there are two distinct concepts of
+"discreteness" which may hold. These are:
+  (i) Every point of `s` is isolated (i.e., the subset topology induced on `s` is the discrete
+      topology).
+ (ii) Every compact subset of `X` meets `s` only finitely often (i.e., the inclusion map `s → X`
+      tends to the cocompact filter along the cofinite filter on `s`).
+
+When `s` is closed, the two conditions are equivalent provided `X` is locally compact and T1,
+see `IsClosed.tendsto_coe_cofinite_iff`. In general however (i) and (ii) are different and it is
+(ii) that is most useful, see e.g., `Subgroup.properlyDiscontinuousSMul_of_tendsto_cofinite`.
```

Comment:
I don't want this to block the PR, but I don't agree with that. I claim that this version is currently more useful because we don't have the right definition of `ProperlyDiscontinuousSMul` (the right one would be in terms of proper maps, although I can't remember it right now), and the one we have is only the right one on locally compact spaces (and we only use it in this context right now in Mathlib). Moreover, if we did have the "right" definition, then `Subgroup.properlyDiscontinuousSMul_of_tendsto_cofinite` would hold *in non locally-compact spaces*, but assuming discreteness in the sense of (i).

Reply:
Interesting: thank you for this comment! I'll reword.

Reply:
I think in an ideal world we would have a class `ProperSMul` saying that the action of topological group on a topological space is proper, and the class `ProperlyDiscontinuousSMul` would be exactly `ProperSMul` when the a-priori-non-topological group is endowed with the discrete topology.